From c4d3e14584fa170dc3e4e4e8fa13b23344de967a Mon Sep 17 00:00:00 2001
From: Thomas Neidhart <tn@apache.org>
Date: Wed, 16 Jan 2013 09:51:52 +0000
Subject: [PATCH] [VFS-285] Keep internal state consistent in case of an error
 of AbstractFileObject.getChildren(). Thanks to Kirill Safonov for the report.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/vfs/trunk@1433872 13f79535-47bb-0310-9956-ffa450edef68
---
 .../apache/commons/vfs2/provider/AbstractFileObject.java   | 7 +++++--
 src/changes/changes.xml                                    | 4 ++++
 2 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/core/src/main/java/org/apache/commons/vfs2/provider/AbstractFileObject.java b/core/src/main/java/org/apache/commons/vfs2/provider/AbstractFileObject.java
index 8632c7399f..576a74d7e0 100644
--- a/core/src/main/java/org/apache/commons/vfs2/provider/AbstractFileObject.java
+++ b/core/src/main/java/org/apache/commons/vfs2/provider/AbstractFileObject.java
@@ -896,14 +896,17 @@ else if (files.length == 0)
             {
                 // Create file objects for the children
                 // children = new FileObject[files.length];
-                children = new FileName[files.length];
+                final FileName[] cache = new FileName[files.length];
                 for (int i = 0; i < files.length; i++)
                 {
                     final String file = files[i];
                     // children[i] = fs.resolveFile(name.resolveName(file, NameScope.CHILD));
                     // children[i] = name.resolveName(file, NameScope.CHILD);
-                    children[i] = fs.getFileSystemManager().resolveName(name, file, NameScope.CHILD);
+                    cache[i] = fs.getFileSystemManager().resolveName(name, file, NameScope.CHILD);
                 }
+                // VFS-285: only assign the children filenames after all of them have been
+                // resolved successfully to prevent an inconsistent internal state
+                children = cache;
             }
 
             return resolveFiles(children);
diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 03da785f8e..29d2f04a9a 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -26,6 +26,10 @@
 <!--       <action issue="VFS-443" dev="ggregory" type="update" due-to="nickallen"> -->
 <!--     	[Local] Need an easy way to convert from a FileObject to a File. -->
 <!--       </action> -->
+      <action issue="VFS-285" dev="tn" type="fix" due-to="Kirill Safonov">
+        AbstractFileObject.getChildren() may corrupt its internal state if a filename
+        can not be resolved.
+      </action>
       <action issue="VFS-451" dev="ggregory" type="fix" due-to="ilmarcoronchi">
         Authentication fails using private key.
       </action>
